internal/executor/executor_test.go

```diff
@@ -79,12 +114,14 @@ func executorTestCases() []executorTestCase {
 				env := []string{"foo=bar", "bat=baz"}
 				exec.SetEnv(env)
 				assert.Equal(t, env, exec.Env())
-				stdout := &bytes.Buffer{}
-				exec.SetStdout(stdout)
+				pipeReader, writer := makePipeReader()
+				pipeReader.start(t)
+				exec.SetStdout(writer)
 				require.NoError(t, exec.Start())
 				require.NoError(t, exec.Wait())
+				pipeReader.closeAndWait()
 				for _, envVar := range env {
-					assert.Contains(t, stdout.String(), envVar)
+					assert.Contains(t, pipeReader.buffer.String(), envVar)
```

Comment:
My instinct is that, despite the logic that Docker shouldn't be writing to this file anymore, the race detector would potentially flag this as a race to read/write the buffer (I've debugged a test that did almost this exact thing, but it was using the built-in `os/exec` library). Does the `buffer` need a rw lock on it?

Reply:
The reason I ended up switching to the pipe reader/writer setup was actually specifically to fix a detected data race. Before there was a Goroutine in runIOStream that would directly write to the buffers read/write interface, and when we read from it it detected a data race. 

In the new setup, pipeReader uses an io.Pipe(), which is safe (and intended) to have a separate goroutine on the reader/writer end, and makes a local goroutine to consume it and put it in the shared buffer. It also has its own sync.WaitGroup to avoid just replicating the data race elsewhere. The test calls closeAndWait which calls Wait() on the WaitGroup, so the goroutine made by pipeReader needs to terminate and call Done() on its waitGroup, avoiding a read/write data race.

The other approach I considered was using the goroutine safe buffer  here: https://gist.github.com/arkan/5924e155dbb4254b64614069ba0afd81 . That is maybe simpler? Would love to hear any thoughts.

Reply:
I could also refactor the code to somehow make it possible to wait on the wait group made by runIOStream.... not sure if that's better?

Reply:
I updated the code to make the wait group used by runIOStream accessible on the object. We now explicitly wait for that wait group.

Reply:
Since that code already exists.... yes let's do that. Thanks for the advice! Will update.

Reply:
Updated